### PR TITLE
Refactor allowEmptyTelemetry to use Items and support hashing

### PR DESCRIPTION
**Pull request by baronfel**
As part of reviewing #30239, I realized that we could answer some questions about the way devs use PublishProfiles from the CLI if we collected a little more telemetry - specifically the name of the profile that was imported (hashed of course), and if a Profile was imported at all. This would allow us to securely get usage rates for SDK-bundled profiles, as well as get data on how often users specify a Profile but there's not a matching profile (i.e. typos or deleted files, that kind of thing).

The Name of the profile would require hashing to comply with out privacy guidelines, since it could contain sensitive user data. The current AllowEmptyTelemetry implementation didn't really allow for specifying if a piece of data needed to be hashed or not, so we needed a new implementation that would allow for this.

The Task now takes in ITaskITems and reads metadata to determine how each one should be handled - `Include` is the key, `Value` is the value, and `Hash` is a boolean that says if the `Value` should be hashed in our standard way.

Tests were updated to ensure that we didn't regress behavior, I added a new one showing the hashing, and I updated existing callsites as part of this.

### CODE REVIEW

File: src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAAllowEmptyTelemetry.cs [hunk 1]
```diff
@@ -8,29 +8,41 @@
...
 
 namespace Microsoft.NET.Build.Tasks.UnitTests
 {
     public class GivenAAllowEmptyTelemetry
     {
-        [Theory]
-        [InlineData("Property1")]
-        [InlineData("Property1=")]
-        public void WhenInvokeWithoutValueItSendValueAsNull(string eventData)
+        private static ITaskItem CreateHashItem(string key, string value = null, Nullable<bool> hash = null) {
```
Comment (nagilson) [inline src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAAllowEmptyTelemetry.cs:18]: I would prefer using string? value and #nullable enable, though this is test code so it's not a big deal.

Comment (nagilson) [reply src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAAllowEmptyTelemetry.cs:18]: However comma, the same could be said about the `var` --> `string?` variables below.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.TargetFrameworkInference.targets [hunk 2]
```diff
@@ -117,7 +117,14 @@ Copyright (c) .NET Foundation. All rights reserved.
   <UsingTask TaskName="AllowEmptyTelemetry" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
 
   <Target Name="_CollectTargetFrameworkForTelemetry" AfterTargets="_CheckForUnsupportedTargetFramework">
-    <AllowEmptyTelemetry EventName="targetframeworkeval" EventData="TargetFrameworkVersion=$([MSBuild]::Escape('$(TargetFrameworkMoniker)'));RuntimeIdentifier=$(RuntimeIdentifier);SelfContained=$(SelfContained);UseApphost=$(UseApphost);OutputType=$(OutputType)" />
+    <ItemGroup>
+      <TFTelemetry Include="TargetFrameworkVersion" Value="$([MSBuild]::Escape('$(TargetFrameworkMoniker)'))" />
```
[the remark this answers is not in the document]
Comment (baronfel) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.TargetFrameworkInference.targets:121]: yeah - I just needed somewhat-unique names for these items. One thing to consider would be to prefix all of the new item types with `_` to signal that they are private.

File: src/Tasks/Microsoft.NET.Build.Tasks/AllowEmptyTelemetry.cs [hunk 3]
```diff
@@ -10,41 +11,67 @@ namespace Microsoft.Build.Tasks
...
         protected override void ExecuteCore()
         {
-            IDictionary<string, string> properties = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
-
-            if (!string.IsNullOrEmpty(EventData))
-            {
-                foreach (string pair in EventData.Split(new[] {';'}, StringSplitOptions.RemoveEmptyEntries))
+            if (EventData == null) {
+                (BuildEngine as IBuildEngine5)?.LogTelemetry(EventName, null);
+            } else {
```
[the remark this answers is not in the document]
Comment (baronfel) [reply src/Tasks/Microsoft.NET.Build.Tasks/AllowEmptyTelemetry.cs:31]: Good point - I haven't applied SDK styling - do we have a way to enforce this?

File: src/Tasks/Microsoft.NET.Build.Tasks/AllowEmptyTelemetry.cs [hunk 4]
```diff
@@ -10,41 +11,67 @@ namespace Microsoft.Build.Tasks
...
+                        properties[key] = value;
                     }
                 }
+                (BuildEngine as IBuildEngine5)?.LogTelemetry(EventName, properties);
             }
+        }
 
-            (BuildEngine as IBuildEngine5)?.LogTelemetry(EventName, properties);
+        // A local copy of Sha256Hasher.HashWithNormalizedCasing from the Microsoft.DotNet.Cli.Utils project.
+        // We don't want to introduce a project<->project dependency, and the logic is straightforward enough.
```
[the remark this answers is not in the document]
Comment (baronfel) [reply src/Tasks/Microsoft.NET.Build.Tasks/AllowEmptyTelemetry.cs:58]: Unsure - this hashing is what most tooling (VS included IIRC) uses for PII, so I _assume_ it's been checked for compliance at some point?

File: src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAAllowEmptyTelemetry.cs (hunk 1 above)
Comment (baronfel) [reply src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAAllowEmptyTelemetry.cs:18]: me too - I wasn't sure what our position on nullability in tests is in the SDK. happy to do it, though.

File: src/Tasks/Microsoft.NET.Build.Tasks/AllowEmptyTelemetry.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (baronfel) [reply src/Tasks/Microsoft.NET.Build.Tasks/AllowEmptyTelemetry.cs:31]: NVM - figured it out.

[the remark this answers is not in the document]
Comment (nagilson) [reply src/Tasks/Microsoft.NET.Build.Tasks/AllowEmptyTelemetry.cs]: I use an extension to auto format files... I think it is this one https://editorconfig.org/ ?

File: src/Tasks/Microsoft.NET.Build.Tasks/AllowEmptyTelemetry.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (nagilson) [reply src/Tasks/Microsoft.NET.Build.Tasks/AllowEmptyTelemetry.cs:63]: Thanks for responding to the feedback, if we can get approval from someone like above then I will be glad to give it a good to merge 👍 I assume it's OK, Sha256 is a reliable algorithm, buuuuuut in the case that this isn't OK, that would be quite bad

[the remark this answers is not in the document]
Comment (baronfel) [reply src/Tasks/Microsoft.NET.Build.Tasks/AllowEmptyTelemetry.cs:63]: @dsplaisted do you know who I should talk to about @nagilson's question?

[the remark this answers is not in the document]
Comment (nagilson) [reply src/Tasks/Microsoft.NET.Build.Tasks/AllowEmptyTelemetry.cs:63]: Got approval from the privacy champ (Daniel) on Teams and discussed with other folks as well to validate.

**Review by nagilson — APPROVED** [review]

**Pull request merged**